src/ralph/api/auth.py

```diff
@@ -0,0 +1,78 @@
+import base64
+import binascii
+import json
+from functools import lru_cache
+
+import bcrypt
+import click
+from fastapi import Request
+from starlette.authentication import (
+    AuthCredentials,
+    AuthenticationBackend,
+    AuthenticationError,
+    SimpleUser,
+)
+
+
+# Unused password used to avoid timing attacks, by comparing passwords supplied with invalid
+# credentials to something innocuous with the same method as if it were a legitimate user.
+UNUSED_PASSWORD = bcrypt.hashpw(b"ralph", bcrypt.gensalt())
+
+
+@lru_cache()
+def get_stored_credentials():
+    """
+    Helper to read the credentials/scopes file to JSON and memoize it so we do not
+    reload it with every request.
+    """
+    auth_file = click.get_current_context().params.get("auth_file")
+    try:
+        stored_credentials = json.load(open(auth_file))
+    except FileNotFoundError as e:
+        raise AuthenticationError(
+            f"Credentials file {auth_file} not found."
+        ) from e
+    return stored_credentials
+
+
+class BasicAuthBackend(AuthenticationBackend):
+    """
+    Custom authentication backend that verifies user credentials passed through basic auth
+    and sets up the user's allowed scopes.
+    """
+
+    async def authenticate(self, request: Request):
+        """
+        Get the basic auth parameters from the Authorization header, and check them against
+        our own list of hashed credentials.
+        """
+        if "Authorization" not in request.headers:
+            raise AuthenticationError("Missing authentication credentials.")
+
+        # Extract basic auth credentials from the relevant header
+        auth = request.headers["Authorization"]
+        try:
+            scheme, credentials = auth.split()
+            if scheme.lower() != "basic":
+                raise AuthenticationError("Missing authentication credentials.")
+            decoded = base64.b64decode(credentials).decode("ascii")
+        except (ValueError, UnicodeDecodeError, binascii.Error) as exc:
+            raise AuthenticationError("Invalid authentication credentials.")
+
+        # Split username/password and check them against our stored user credentials
+        username, _, password = decoded.partition(":")
+        user_info = get_stored_credentials().get(username, {})
```

Comment:
AUTH_FILE should be defined in `ralph.defaults` using the `config` utility:

```python
# ralph/defaults.py
AUTH_FILE = Path(config("RALPH_AUTH_FILE", APP_DIR / "auth.json"))
```

And I would prefer the auth file to be a list of user entries such as:

```json
[
  {"username": "foo"},
  {"username": "bar"}
]
```

```suggestion
        user_info = filter(
            lambda u: u.get("username") == username, 
            get_stored_credentials(AUTH_FILE)
        )
```